File: backend/src/Solitaire.Engine/Klondike.cs
using System.Collections.Immutable;

namespace Solitaire.Engine;

/// <summary>
/// The authoritative, pure Klondike engine: deal, move validation/application,
/// legal-move generation, win detection, and deterministic replay. No method
/// mutates its inputs; every change yields a new <see cref="GameState"/>.
/// </summary>
public static class Klondike
{
    /// <summary>Creates the initial game state for a seed and rule set.</summary>
    public static GameState NewGame(int seed, GameOptions options)
    {
        options.Validate();

        var shuffled = Deck.Shuffle(seed);
        var (tableau, stock) = Deck.Deal(shuffled);

        return new GameState(
            options,
            stock: stock,
            waste: ImmutableArray<Card>.Empty,
            foundations: ImmutableArray.Create(0, 0, 0, 0),
            tableau: tableau,
            score: 0,
            redealsUsed: 0);
    }

    /// <summary>True once the game is won (all foundations complete).</summary>
    public static bool IsWin(GameState state) => state.IsWon;

    /// <summary>
    /// Validates <paramref name="move"/> against the rules of
    /// <paramref name="state"/>. On success returns true and yields the resulting
    /// <paramref name="next"/> state and the nominal <paramref name="scoreDelta"/>.
    /// On failure returns false, sets <paramref name="next"/> to the unchanged
    /// input state, and makes no changes (no mutation on failure).
    /// </summary>
    /// <remarks>
    /// <paramref name="scoreDelta"/> is the nominal points for the move; the
    /// cumulative <see cref="GameState.Score"/> is clamped so it never goes below
    /// zero, so a large penalty may change the score by less than the delta.
    /// </remarks>
    public static bool TryApplyMove(GameState state, Move move, out GameState next, out int scoreDelta)
    {
        ArgumentNullException.ThrowIfNull(state);

        var result = Apply(state, move);
        if (result is null)
        {
            next = state;
            scoreDelta = 0;
            return false;
        }

        var (partial, delta) = result.Value;
        scoreDelta = delta;
        next = partial.With(score: Scoring.Clamp(state.Score + delta));
        return true;
    }

    /// <summary>
    /// Returns every legal move from <paramref name="state"/> (used for hints).
    /// </summary>
    /// <remarks>
    /// One pruning rule keeps the list useful: moving an entire face-up-only
    /// column (no face-down cards beneath) onto an empty column is omitted because
    /// it relocates a King without exposing anything. Such a move is still
    /// <em>accepted</em> by <see cref="TryApplyMove"/>; it is merely not suggested.
    /// </remarks>
    public static IReadOnlyList<Move> GetLegalMoves(GameState state)
    {
        ArgumentNullException.ThrowIfNull(state);

        var moves = new List<Move>();

        // Stock / waste cycling.
        if (!state.Stock.IsEmpty)
        {
            moves.Add(Move.Draw());
        }
        else if (!state.Waste.IsEmpty && state.RedealsUsed < state.Options.MaxRedeals)
        {
            moves.Add(Move.Recycle());
        }

        // Waste top plays.
        if (state.WasteTop is { } wasteCard)
        {
            if (CanPlaceOnFoundation(state, wasteCard))
            {
                moves.Add(Move.WasteToFoundation());
            }

            for (int t = 0; t < GameState.TableauCount; t++)
            {
                if (CanPlaceOnTableau(state.Tableau[t], wasteCard))
                {
                    moves.Add(Move.WasteToTableau(t));
                }
            }
        }

        // Tableau top -> foundation.
        for (int s = 0; s < GameState.TableauCount; s++)
        {
            if (state.Tableau[s].TopCard is { } top && CanPlaceOnFoundation(state, top))
            {
                moves.Add(Move.TableauToFoundation(s));
            }
        }

        // Foundation -> tableau.
        for (int f = 0; f < GameState.FoundationCount; f++)
        {
            int rank = state.Foundations[f];
            if (rank == 0)
            {
                continue;
            }

            var card = new Card((Suit)f, rank);
            for (int t = 0; t < GameState.TableauCount; t++)
            {
                if (CanPlaceOnTableau(state.Tableau[t], card))
                {
                    moves.Add(Move.FoundationToTableau((Suit)f, t));
                }
            }
        }

        // Tableau -> tableau (every legal run length).
        for (int s = 0; s < GameState.TableauCount; s++)
        {
            var source = state.Tableau[s];
            for (int count = 1; count <= source.FaceUpCount; count++)
            {
                var bottom = source.Cards[source.Count - count];
                for (int t = 0; t < GameState.TableauCount; t++)
                {
                    if (t == s)
                    {
                        continue;
                    }

                    var dest = state.Tableau[t];

                    // Skip relocating a whole face-up-only column onto an empty one.
                    if (dest.IsEmpty
                        && source.FaceDownCount == 0
                        && count == source.FaceUpCount)
                    {
                        continue;
                    }

                    if (CanPlaceOnTableau(dest, bottom))
                    {
                        moves.Add(Move.TableauToTableau(s, t, count));
                    }
                }
            }
        }

        return moves;
    }

    /// <summary>
    /// Replays a move sequence from a fresh game. Pure and side-effect free: the
    /// result reports whether every move was legal, the final state and score, and
    /// whether the game was completed as a win. Any illegal move stops the replay
    /// and is reported via <see cref="ReplayResult.FirstIllegalMoveIndex"/> — this
    /// is the anti-cheat guarantee.
    /// </summary>
    public static ReplayResult<GameState> Replay(int seed, GameOptions options, IReadOnlyList<Move> moves)
    {
        ArgumentNullException.ThrowIfNull(moves);

        var state = NewGame(seed, options);

        for (int i = 0; i < moves.Count; i++)
        {
            if (!TryApplyMove(state, moves[i], out var nextState, out _))
            {
                return new ReplayResult<GameState>(
                    FinalState: state,
                    Score: state.Score,
                    Won: false,
                    AllMovesLegal: false,
                    FirstIllegalMoveIndex: i);
            }

            state = nextState;
        }

        return new ReplayResult<GameState>(
            FinalState: state,
            Score: state.Score,
            Won: state.IsWon,
            AllMovesLegal: true,
            FirstIllegalMoveIndex: null);
    }

    // ----------------------------------------------------------------------
    // Portable move (de)serialization for vectors and the API.
    // ----------------------------------------------------------------------

    /// <summary>Converts a Klondike move to its portable <see cref="MoveDto"/> form.</summary>
    public static MoveDto ToDto(Move move) => move.Type switch
    {
        MoveType.Draw => new MoveDto("Draw"),
        MoveType.Recycle => new MoveDto("Recycle"),
        MoveType.WasteToFoundation => new MoveDto("WasteToFoundation"),
        MoveType.WasteToTableau => new MoveDto("WasteToTableau", Destination: move.Destination),
        MoveType.TableauToFoundation => new MoveDto("TableauToFoundation", Source: move.Source),
        MoveType.FoundationToTableau =>
            new MoveDto("FoundationToTableau", Source: move.Source, Destination: move.Destination),
        MoveType.TableauToTableau => new MoveDto(
            "TableauToTableau", Source: move.Source, Destination: move.Destination, Count: move.Count),
        _ => throw new ArgumentOutOfRangeException(nameof(move), move.Type, "Unknown move type."),
    };

    /// <summary>Parses a portable <see cref="MoveDto"/> into a Klondike move.</summary>
    public static Move FromDto(MoveDto dto)
    {
        ArgumentNullException.ThrowIfNull(dto);
        return dto.Type switch
        {
            "Draw" => Move.Draw(),
            "Recycle" => Move.Recycle(),
            "WasteToFoundation" => Move.WasteToFoundation(),
            "WasteToTableau" => Move.WasteToTableau(Req(dto.Destination, "destination")),
            "TableauToFoundation" => Move.TableauToFoundation(Req(dto.Source, "source")),
            "FoundationToTableau" =>
                Move.FoundationToTableau((Suit)Req(dto.Source, "source"), Req(dto.Destination, "destination")),
            "TableauToTableau" => Move.TableauToTableau(
                Req(dto.Source, "source"), Req(dto.Destination, "destination"), Req(dto.Count, "count")),
            _ => throw new ArgumentOutOfRangeException(nameof(dto), dto.Type, "Unknown move type."),
        };
    }

    private static int Req(int? value, string field) =>
        value ?? throw new ArgumentException($"Klondike move is missing '{field}'.");

    /// <summary>Reads Klondike options from a portable options bag.</summary>
    public static GameOptions OptionsFromBag(IReadOnlyDictionary<string, int> options)
    {
        ArgumentNullException.ThrowIfNull(options);
        if (!options.TryGetValue("drawCount", out int drawCount))
        {
            throw new ArgumentException("Klondike options require 'drawCount'.");
        }

        int maxRedeals = options.TryGetValue("maxRedeals", out int mr) ? mr : GameOptions.Unlimited;
        var result = new GameOptions(drawCount, maxRedeals);
        result.Validate();
        return result;
    }

    // ----------------------------------------------------------------------
    // Rule helpers
    // ----------------------------------------------------------------------

    /// <summary>Can <paramref name="moving"/> be placed on foundation of its suit?</summary>
    private static bool CanPlaceOnFoundation(GameState state, Card moving) =>
        state.Foundations[(int)moving.Suit] == moving.Rank - 1;

    /// <summary>
    /// Can <paramref name="moving"/> (the bottom card of the moved run) be placed
    /// on <paramref name="pile"/>? Kings on empty piles; otherwise one rank below
    /// the top card and of the opposite color.
    /// </summary>
    private static bool CanPlaceOnTableau(TableauPile pile, Card moving)
    {
        if (pile.IsEmpty)
        {
            return moving.Rank == Card.King;
        }

        Card top = pile.Cards[^1];
        return moving.Rank == top.Rank - 1 && moving.Color != top.Color;
    }

    /// <summary>True if the face-up cards <paramref name="run"/> form an alternating,
    /// strictly-descending sequence from bottom (index 0) to top.</summary>
    private static bool IsValidRun(ReadOnlySpan<Card> run)
    {
        for (int i = 1; i < run.Length; i++)
        {
            Card lower = run[i];
            Card upper = run[i - 1];
            if (lower.Rank != upper.Rank - 1 || lower.Color == upper.Color)
            {
                return false;
            }
        }

        return true;
    }

    // ----------------------------------------------------------------------
    // Move application. Returns null on any rule violation. Never mutates input.
    // The returned state carries the *unclamped* running score is left to the
    // caller; here we only compute the state delta and the nominal points.
    // ----------------------------------------------------------------------

    private static (GameState State, int ScoreDelta)? Apply(GameState state, Move move) => move.Type switch
    {
        MoveType.Draw => ApplyDraw(state),
        MoveType.Recycle => ApplyRecycle(state),
        MoveType.WasteToFoundation => ApplyWasteToFoundation(state),
        MoveType.WasteToTableau => ApplyWasteToTableau(state, move),
        MoveType.TableauToTableau => ApplyTableauToTableau(state, move),
        MoveType.TableauToFoundation => ApplyTableauToFoundation(state, move),
        MoveType.FoundationToTableau => ApplyFoundationToTableau(state, move),
        _ => null,
    };

    private static (GameState, int)? ApplyDraw(GameState state)
    {
        if (state.Stock.IsEmpty)
        {
            return null;
        }

        int k = Math.Min(state.Options.DrawCount, state.Stock.Length);
        var drawn = state.Stock.AsSpan(0, k);
        var newStock = state.Stock.RemoveRange(0, k);
        var newWaste = state.Waste.AddRange(drawn);

        return (state.With(stock: newStock, waste: newWaste), 0);
    }

    private static (GameState, int)? ApplyRecycle(GameState state)
    {
        if (!state.Stock.IsEmpty || state.Waste.IsEmpty)
        {
            return null;
        }

        if (state.RedealsUsed >= state.Options.MaxRedeals)
        {
            return null;
        }

        // Flipping the waste pile over restores the original draw order:
        // waste[0] (first drawn) becomes the new stock top.
        var newStock = state.Waste;
        int penalty = state.Options.DrawCount == 1 ? Scoring.RecycleDrawOne : Scoring.RecycleDrawThree;

        return (
            state.With(
                stock: newStock,
                waste: ImmutableArray<Card>.Empty,
                redealsUsed: state.RedealsUsed + 1),
            penalty);
    }

    private static (GameState, int)? ApplyWasteToFoundation(GameState state)
    {
        if (state.WasteTop is not { } card || !CanPlaceOnFoundation(state, card))
        {
            return null;
        }

        var newFoundations = state.Foundations.SetItem((int)card.Suit, card.Rank);
        var newWaste = state.Waste.RemoveAt(state.Waste.Length - 1);

        return (state.With(waste: newWaste, foundations: newFoundations), Scoring.WasteToFoundation);
    }

    private static (GameState, int)? ApplyWasteToTableau(GameState state, Move move)
    {
        if (!IsTableauIndex(move.Destination) || state.WasteTop is not { } card)
        {
            return null;
        }

        var dest = state.Tableau[move.Destination];
        if (!CanPlaceOnTableau(dest, card))
        {
            return null;
        }

        var newWaste = state.Waste.RemoveAt(state.Waste.Length - 1);
        var newTableau = state.Tableau.SetItem(move.Destination, dest.Append([card]));

        return (state.With(waste: newWaste, tableau: newTableau), Scoring.WasteToTableau);
    }

    private static (GameState, int)? ApplyTableauToFoundation(GameState state, Move move)
    {
        if (!IsTableauIndex(move.Source))
        {
            return null;
        }

        var source = state.Tableau[move.Source];
        if (source.TopCard is not { } card || !CanPlaceOnFoundation(state, card))
        {
            return null;
        }

        var newSource = source.RemoveTop(1, out bool flipped);
        var newFoundations = state.Foundations.SetItem((int)card.Suit, card.Rank);
        var newTableau = state.Tableau.SetItem(move.Source, newSource);

        int delta = Scoring.TableauToFoundation + (flipped ? Scoring.TurnOverTableauCard : 0);
        return (state.With(tableau: newTableau, foundations: newFoundations), delta);
    }

    private static (GameState, int)? ApplyFoundationToTableau(GameState state, Move move)
    {
        if (move.Source is < 0 or >= GameState.FoundationCount || !IsTableauIndex(move.Destination))
        {
            return null;
        }

        int rank = state.Foundations[move.Source];
        if (rank == 0)
        {
            return null;
        }

        var card = new Card((Suit)move.Source, rank);
        var dest = state.Tableau[move.Destination];
        if (!CanPlaceOnTableau(dest, card))
        {
            return null;
        }

        var newFoundations = state.Foundations.SetItem(move.Source, rank - 1);
        var newTableau = state.Tableau.SetItem(move.Destination, dest.Append([card]));

        return (state.With(tableau: newTableau, foundations: newFoundations), Scoring.FoundationToTableau);
    }

    private static (GameState, int)? ApplyTableauToTableau(GameState state, Move move)
    {
        if (!IsTableauIndex(move.Source) || !IsTableauIndex(move.Destination) || move.Source == move.Destination)
        {
            return null;
        }

        var source = state.Tableau[move.Source];
        if (move.Count < 1 || move.Count > source.FaceUpCount)
        {
            return null;
        }

        var moved = source.Cards.AsSpan(source.Count - move.Count, move.Count);
        if (!IsValidRun(moved))
        {
            return null;
        }

        var dest = state.Tableau[move.Destination];
        if (!CanPlaceOnTableau(dest, moved[0]))
        {
            return null;
        }

        var newSource = source.RemoveTop(move.Count, out bool flipped);
        var newDest = dest.Append(moved);
        var newTableau = state.Tableau
            .SetItem(move.Source, newSource)
            .SetItem(move.Destination, newDest);

        int delta = flipped ? Scoring.TurnOverTableauCard : 0;
        return (state.With(tableau: newTableau), delta);
    }

    private static bool IsTableauIndex(int index) => index is >= 0 and < GameState.TableauCount;
}

File: frontend/src/engine/klondike.ts
// Klondike engine — a behavioral port of Klondike.cs / Scoring.cs / GameState.cs.
// Same deal, same move legality, same scoring, same clamping.

import { KING, sameColor, type Card } from './cards';
import { dealKlondike, KLONDIKE_TABLEAU_COUNT, shuffle } from './deck';
import {
  appendCards,
  faceUpCount,
  isPileEmpty,
  removeTop,
  topCard,
  type TableauPile,
} from './tableau';
import type {
  ApplyResult,
  GameDefinition,
  MoveDto,
  ReplayOutcome,
  ReplayResult,
  SolitaireEngine,
} from './types';

const FOUNDATION_COUNT = 4;
export const KLONDIKE_UNLIMITED_REDEALS = 2147483647; // int.MaxValue

// Documented scoring model (Scoring.cs).
const SCORE = {
  wasteToFoundation: 10,
  tableauToFoundation: 10,
  wasteToTableau: 5,
  turnOver: 5,
  foundationToTableau: -15,
  recycleDraw1: -100,
  recycleDraw3: -20,
} as const;

const clamp = (score: number): number => (score < 0 ? 0 : score);

export interface KlondikeOptions {
  readonly drawCount: number; // 1 or 3
  readonly maxRedeals: number; // KLONDIKE_UNLIMITED_REDEALS == unlimited
}

export interface KlondikeState {
  readonly options: KlondikeOptions;
  readonly stock: readonly Card[]; // index 0 = top (next drawn)
  readonly waste: readonly Card[]; // last element = playable top
  readonly foundations: readonly number[]; // top rank per suit index (0 = empty)
  readonly tableau: readonly TableauPile[];
  readonly score: number;
  readonly redealsUsed: number;
}

export function klondikeIsWon(state: KlondikeState): boolean {
  return state.foundations.every((top) => top === KING);
}

export function klondikeNewGame(seed: number, options: KlondikeOptions): KlondikeState {
  validateOptions(options);
  const { tableau, stock } = dealKlondike(shuffle(seed));
  return {
    options,
    stock,
    waste: [],
    foundations: [0, 0, 0, 0],
    tableau,
    score: 0,
    redealsUsed: 0,
  };
}

type Applied = { readonly state: KlondikeState; readonly delta: number } | null;

export function klondikeTryApplyMove(
  state: KlondikeState,
  move: MoveDto,
): ApplyResult<KlondikeState> {
  const result = apply(state, move);
  if (result === null) {
    return { ok: false, next: state, scoreDelta: 0 };
  }
  return {
    ok: true,
    next: { ...result.state, score: clamp(state.score + result.delta) },
    scoreDelta: result.delta,
  };
}

export function klondikeReplay(
  seed: number,
  options: KlondikeOptions,
  moves: readonly MoveDto[],
): ReplayResult<KlondikeState> {
  let state = klondikeNewGame(seed, options);
  for (let i = 0; i < moves.length; i++) {
    const applied = klondikeTryApplyMove(state, moves[i]!);
    if (!applied.ok) {
      return {
        finalState: state,
        score: state.score,
        won: false,
        allMovesLegal: false,
        firstIllegalMoveIndex: i,
      };
    }
    state = applied.next;
  }
  return {
    finalState: state,
    score: state.score,
    won: klondikeIsWon(state),
    allMovesLegal: true,
    firstIllegalMoveIndex: null,
  };
}

export function klondikeGetLegalMoves(state: KlondikeState): MoveDto[] {
  const moves: MoveDto[] = [];

  if (state.stock.length > 0) {
    moves.push({ type: 'Draw' });
  } else if (state.waste.length > 0 && state.redealsUsed < state.options.maxRedeals) {
    moves.push({ type: 'Recycle' });
  }

  const wasteCard = wasteTop(state);
  if (wasteCard !== undefined) {
    if (canPlaceOnFoundation(state, wasteCard)) {
      moves.push({ type: 'WasteToFoundation' });
    }
    for (let t = 0; t < KLONDIKE_TABLEAU_COUNT; t++) {
      if (canPlaceOnTableau(state.tableau[t]!, wasteCard)) {
        moves.push({ type: 'WasteToTableau', destination: t });
      }
    }
  }

  for (let s = 0; s < KLONDIKE_TABLEAU_COUNT; s++) {
    const top = topCard(state.tableau[s]!);
    if (top !== undefined && canPlaceOnFoundation(state, top)) {
      moves.push({ type: 'TableauToFoundation', source: s });
    }
  }

  for (let f = 0; f < FOUNDATION_COUNT; f++) {
    const rank = state.foundations[f]!;
    if (rank === 0) {
      continue;
    }
    const card: Card = { suit: f, rank };
    for (let t = 0; t < KLONDIKE_TABLEAU_COUNT; t++) {
      if (canPlaceOnTableau(state.tableau[t]!, card)) {
        moves.push({ type: 'FoundationToTableau', source: f, destination: t });
      }
    }
  }

  for (let s = 0; s < KLONDIKE_TABLEAU_COUNT; s++) {
    const source = state.tableau[s]!;
    const up = faceUpCount(source);
    for (let count = 1; count <= up; count++) {
      const bottom = source.cards[source.cards.length - count]!;
      for (let t = 0; t < KLONDIKE_TABLEAU_COUNT; t++) {
        if (t === s) {
          continue;
        }
        const dest = state.tableau[t]!;
        if (isPileEmpty(dest) && source.faceDownCount === 0 && count === up) {
          continue; // futile whole-column relocation
        }
        if (canPlaceOnTableau(dest, bottom)) {
          moves.push({ type: 'TableauToTableau', source: s, destination: t, count });
        }
      }
    }
  }

  return moves;
}

export function klondikeOptionsFromBag(bag: Readonly<Record<string, number>>): KlondikeOptions {
  const drawCount = bag['drawCount'];
  if (drawCount === undefined) {
    throw new Error("Klondike options require 'drawCount'.");
  }
  const maxRedeals = bag['maxRedeals'] ?? KLONDIKE_UNLIMITED_REDEALS;
  const options = { drawCount, maxRedeals };
  validateOptions(options);
  return options;
}

export const klondikeEngine: SolitaireEngine = {
  variant: 'klondike',
  replay(game: GameDefinition): ReplayOutcome {
    const options = klondikeOptionsFromBag(game.options);
    const result = klondikeReplay(game.seed, options, game.moves);
    return {
      score: result.score,
      won: result.won,
      allMovesLegal: result.allMovesLegal,
      firstIllegalMoveIndex: result.firstIllegalMoveIndex,
    };
  },
};

// -- rule helpers -------------------------------------------------------------

function validateOptions(options: KlondikeOptions): void {
  if (options.drawCount !== 1 && options.drawCount !== 3) {
    throw new Error('DrawCount must be 1 or 3.');
  }
  if (options.maxRedeals < 0) {
    throw new Error('MaxRedeals must be >= 0.');
  }
}

function wasteTop(state: KlondikeState): Card | undefined {
  return state.waste.length === 0 ? undefined : state.waste[state.waste.length - 1];
}

function canPlaceOnFoundation(state: KlondikeState, moving: Card): boolean {
  return state.foundations[moving.suit] === moving.rank - 1;
}

function canPlaceOnTableau(pile: TableauPile, moving: Card): boolean {
  if (isPileEmpty(pile)) {
    return moving.rank === KING;
  }
  const top = pile.cards[pile.cards.length - 1]!;
  return moving.rank === top.rank - 1 && !sameColor(moving.suit, top.suit);
}

function isValidRun(run: readonly Card[]): boolean {
  for (let i = 1; i < run.length; i++) {
    const lower = run[i]!;
    const upper = run[i - 1]!;
    if (lower.rank !== upper.rank - 1 || sameColor(lower.suit, upper.suit)) {
      return false;
    }
  }
  return true;
}

function isTableauIndex(index: number): boolean {
  return index >= 0 && index < KLONDIKE_TABLEAU_COUNT;
}

// -- move application (returns null on any rule violation) --------------------

function apply(state: KlondikeState, move: MoveDto): Applied {
  switch (move.type) {
    case 'Draw':
      return applyDraw(state);
    case 'Recycle':
      return applyRecycle(state);
    case 'WasteToFoundation':
      return applyWasteToFoundation(state);
    case 'WasteToTableau':
      return applyWasteToTableau(state, move.destination ?? -1);
    case 'TableauToFoundation':
      return applyTableauToFoundation(state, move.source ?? -1);
    case 'FoundationToTableau':
      return applyFoundationToTableau(state, move.source ?? -1, move.destination ?? -1);
    case 'TableauToTableau':
      return applyTableauToTableau(
        state,
        move.source ?? -1,
        move.destination ?? -1,
        move.count ?? 0,
      );
    default:
      return null;
  }
}

function applyDraw(state: KlondikeState): Applied {
  if (state.stock.length === 0) {
    return null;
  }
  const k = Math.min(state.options.drawCount, state.stock.length);
  const drawn = state.stock.slice(0, k);
  return {
    state: { ...state, stock: state.stock.slice(k), waste: [...state.waste, ...drawn] },
    delta: 0,
  };
}

function applyRecycle(state: KlondikeState): Applied {
  if (state.stock.length > 0 || state.waste.length === 0) {
    return null;
  }
  if (state.redealsUsed >= state.options.maxRedeals) {
    return null;
  }
  // Flipping the waste over restores the original draw order: waste[0] (first
  // drawn) becomes the new stock top.
  const penalty = state.options.drawCount === 1 ? SCORE.recycleDraw1 : SCORE.recycleDraw3;
  return {
    state: { ...state, stock: [...state.waste], waste: [], redealsUsed: state.redealsUsed + 1 },
    delta: penalty,
  };
}

function applyWasteToFoundation(state: KlondikeState): Applied {
  const card = wasteTop(state);
  if (card === undefined || !canPlaceOnFoundation(state, card)) {
    return null;
  }
  const foundations = [...state.foundations];
  foundations[card.suit] = card.rank;
  return {
    state: { ...state, foundations, waste: state.waste.slice(0, -1) },
    delta: SCORE.wasteToFoundation,
  };
}

function applyWasteToTableau(state: KlondikeState, destination: number): Applied {
  const card = wasteTop(state);
  if (!isTableauIndex(destination) || card === undefined) {
    return null;
  }
  const dest = state.tableau[destination]!;
  if (!canPlaceOnTableau(dest, card)) {
    return null;
  }
  const tableau = [...state.tableau];
  tableau[destination] = appendCards(dest, [card]);
  return {
    state: { ...state, waste: state.waste.slice(0, -1), tableau },
    delta: SCORE.wasteToTableau,
  };
}

function applyTableauToFoundation(state: KlondikeState, source: number): Applied {
  if (!isTableauIndex(source)) {
    return null;
  }
  const pile = state.tableau[source]!;
  const card = topCard(pile);
  if (card === undefined || !canPlaceOnFoundation(state, card)) {
    return null;
  }
  const { pile: newSource, flipped } = removeTop(pile, 1);
  const foundations = [...state.foundations];
  foundations[card.suit] = card.rank;
  const tableau = [...state.tableau];
  tableau[source] = newSource;
  return {
    state: { ...state, foundations, tableau },
    delta: SCORE.tableauToFoundation + (flipped ? SCORE.turnOver : 0),
  };
}

function applyFoundationToTableau(
  state: KlondikeState,
  source: number,
  destination: number,
): Applied {
  if (source < 0 || source >= FOUNDATION_COUNT || !isTableauIndex(destination)) {
    return null;
  }
  const rank = state.foundations[source]!;
  if (rank === 0) {
    return null;
  }
  const card: Card = { suit: source, rank };
  const dest = state.tableau[destination]!;
  if (!canPlaceOnTableau(dest, card)) {
    return null;
  }
  const foundations = [...state.foundations];
  foundations[source] = rank - 1;
  const tableau = [...state.tableau];
  tableau[destination] = appendCards(dest, [card]);
  return { state: { ...state, foundations, tableau }, delta: SCORE.foundationToTableau };
}

function applyTableauToTableau(
  state: KlondikeState,
  source: number,
  destination: number,
  count: number,
): Applied {
  if (!isTableauIndex(source) || !isTableauIndex(destination) || source === destination) {
    return null;
  }
  const src = state.tableau[source]!;
  if (count < 1 || count > faceUpCount(src)) {
    return null;
  }
  const moved = src.cards.slice(src.cards.length - count);
  if (!isValidRun(moved)) {
    return null;
  }
  const dest = state.tableau[destination]!;
  if (!canPlaceOnTableau(dest, moved[0]!)) {
    return null;
  }
  const { pile: newSource, flipped } = removeTop(src, count);
  const tableau = [...state.tableau];
  tableau[source] = newSource;
  tableau[destination] = appendCards(dest, moved);
  return { state: { ...state, tableau }, delta: flipped ? SCORE.turnOver : 0 };
}
